core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,555 @@
+
+  private ClientStreamListener masterListener;
+
+  RetriableStream(
+      MethodDescriptor<ReqT, ?> method,
+      CallOptions callOptions,
+      Metadata headers,
+      ClientTransportProvider clientTransportProvider,
+      Context context) {
+    this.method = method;
+    this.callOptions = callOptions;
+    this.headers = headers;
+    this.clientTransportProvider = clientTransportProvider;
+    delayedClientTransport = clientTransportProvider.getDelayedTransport();
+    this.context = context;
+  }
+
+  private boolean commit(ClientStream winningSubstream) {
+    delayedClientTransport.removeUncommittedRetriableStream(this);
+
+    Collection<ClientStream> savedDrainedSubstreams;
+    synchronized (lock) {
+      if (state.winningSubstream != null) {
+        return false;
+      }
+
+      savedDrainedSubstreams = state.drainedSubstreams;
+
+      state = state.committed(winningSubstream);
+    }
+
+    // For hedging only, not needed for normal retry
+    // TODO(zdapeng): also cancel all the scheduled hedges.
+    for (ClientStream substream : savedDrainedSubstreams) {
+      if (substream != winningSubstream) {
+        substream.cancel(Status.CANCELLED.withDescription(CANCELLED_BECAUSE_HEDGE_COMMITTED));
+      }
+    }
+
+    return true;
+  }
+
+  private void retry() {
+    ClientStream substream = newSubstream();
+
+    // TODO(zdapeng): update "grpc-retry-attempts" header
+
+    drain(substream);
+  }
+
+  private ClientStream newSubstream() {
+    ClientTransport transport =
+        clientTransportProvider.get(new PickSubchannelArgsImpl(method, headers, callOptions));
+    Context origContext = context.attach();
+    try {
+      return transport.newStream(method, headers, callOptions);
+    } finally {
+      context.detach(origContext);
+    }
+  }
+
+  private void drain(ClientStream substream) {
+    int index = 0;
+    int chunk = 0x80;
+    List<BufferEntry> list = null;
+    Status status;
+
+    while (true) {
+      synchronized (lock) {
+        int stop = Math.min(index + chunk, state.buffer.size());
+        if (list == null) {
+          list = new ArrayList<BufferEntry>(stop - index);
+        }
+        list.clear();
+        list.addAll(state.buffer.subList(index, stop));
+        index = stop;
+      }
+
+      for (BufferEntry bufferEntry : list) {
+        State savedState = state;
+        if (savedState.cancellingStatus != null) {
+          substream.cancel(savedState.cancellingStatus);
+          return;
+        }
+        if (savedState.winningSubstream != null && savedState.winningSubstream != substream) {
+          // committed but not me
+          substream.cancel(Status.CANCELLED.withDescription(CANCELLED_BECAUSE_HEDGE_COMMITTED));
+          return;
+        }
+        bufferEntry.runWith(substream);
+      }
+
+      synchronized (lock) {
+        if (index != state.buffer.size()) { // I'm not drained
```

Comment:
Re-load savedState and use it.

I actually think this synchronized block can integrated with the earlier initial synchronized block.